internal/emigration/emigrate-nobles.lua

```diff
@@ -0,0 +1,347 @@
+--@module = true
+
+--[[
+TODO:
+  * Feature: have rightful ruler immigrate to fort if off-site
+  * Noble of other civ residing in fort?
+]]--
+
+local argparse = require("argparse")
+
+local unit_link_utils = reqscript("internal/emigration/unit-link-utils")
+
+local options = {
+    all = false,
+    unitId = -1,
+    list = false
+}
+
+-- adapted from Units::get_land_title()
+---@return df.world_site|nil
+local function findSiteOfRule(np)
+    local site = nil
+    local civ = np.entity -- lawmakers seem to be all civ-level positions
+    for _, link in ipairs(civ.site_links) do
+        if not link.flags.land_for_holding then goto continue end
+        if link.position_profile_id ~= np.assignment.id then goto continue end
+
+        site = df.world_site.find(link.target)
+        break
+        ::continue::
+    end
+
+    return site
+end
+
+---@return df.world_site|nil
+local function findCapital(civ)
+    local civCapital = nil
+    for _, link in ipairs(civ.site_links) do
+        if link.flags.capital then
+            civCapital = df.world_site.find(link.target)
+            break
+        end
+    end
+
+    return civCapital
+end
+
+---@param unit df.unit
+---@param nobleList { unit: df.unit, site: df.world_site, id: number }[]
+---@param thisSite df.world_site
+---@param civ df.historical_entity
+local function addNobleOfOtherSite(unit, nobleList, thisSite, civ)
+    local nps = dfhack.units.getNoblePositions(unit) or {}
+    local noblePos = nil
+    for _, np in ipairs(nps) do
+        if np.position.flags.IS_LAW_MAKER then
+            noblePos = np
+            break
+        end
+    end
+
+    if not noblePos then return end -- unit is not nobility
+
+    -- Monarchs do not seem to have an world_site associated to them (?)
+    if noblePos.position.flags.RULES_FROM_LOCATION and noblePos.entity.id == civ.id then
+        local capital = findCapital(civ)
+        if capital and capital.id ~= thisSite.id then
+            table.insert(nobleList, {unit = unit, site = capital, id = noblePos.assignment.id})
+        end
+        return
+    end
+
+    local name = dfhack.units.getReadableName(unit)
+    -- Logic for dukes, counts, barons
+    local site = findSiteOfRule(noblePos)
+    if not site then qerror("could not find land of "..name) end
+
+    if site.id == thisSite.id then return end -- noble rules current fort
+    table.insert(nobleList, {unit = unit, site = site, id = noblePos.assignment.id})
+end
+
+---@param unit df.unit
+local function removeMandates(unit)
+    local mandates = df.global.world.mandates
```

Comment:
Could you update to the latest code? If you're not building the DFHack binaries, you can pick up the latest build by subscribing to the DFHack `testing` branch on Steam. This should be changed to `df.global.world.mandates.all`, and there may be other types/paths/field names that need to be updated according to the just-merged Great Reorganization

Reply:
Yep! I'll do another pass tomorrow, realised there are still some loose ends yet.